quickjs.c

```diff
@@ -51507,3 +51631,166 @@ uint32_t add_ic_slot1(JSContext *ctx, JSInlineCache *ic, JSAtom atom)
 end:
     return 0;
 }
+
+/* CallSite */
+
+static void js_callsite_finalizer(JSRuntime *rt, JSValue val)
+{
+    JSCallSiteData *csd = JS_GetOpaque(val, JS_CLASS_CALL_SITE);
+    if (csd) {
+        JS_FreeValueRT(rt, csd->filename);
+        JS_FreeValueRT(rt, csd->func);
+        JS_FreeValueRT(rt, csd->func_name);
+        js_free_rt(rt, csd);
+    }
+}
+
+static void js_callsite_mark(JSRuntime *rt, JSValue val, JS_MarkFunc *mark_func)
+{
+    JSCallSiteData *csd = JS_GetOpaque(val, JS_CLASS_CALL_SITE);
+    if (csd) {
+        JS_MarkValue(rt, csd->filename, mark_func);
+        JS_MarkValue(rt, csd->func, mark_func);
+        JS_MarkValue(rt, csd->func_name, mark_func);
+    }
+}
+
+static JSValue js_new_callsite(JSContext *ctx, JSCallSiteData *csd) {
+    JSValue obj = js_create_from_ctor(ctx, JS_UNDEFINED, JS_CLASS_CALL_SITE);
+    if (JS_IsException(obj))
+        return JS_EXCEPTION;
+
+    JSCallSiteData *csd1 = js_malloc(ctx, sizeof(*csd));
+    if (!csd1) {
+        JS_FreeValue(ctx, obj);
+        return JS_EXCEPTION;
+    }
+
+    memcpy(csd1, csd, sizeof(*csd));
+
+    JS_SetOpaque(obj, csd1);
+
+    return obj;
+}
+
+static void js_new_callsite_data(JSContext *ctx, JSCallSiteData *csd, JSStackFrame *sf)
+{
+    const char *func_name_str;
+    JSObject *p;
+
+    csd->func = js_dup(sf->cur_func);
+    func_name_str = get_func_name(ctx, sf->cur_func);
+    if (!func_name_str || func_name_str[0] == '\0')
+        csd->func_name = JS_NULL;
+    else
+        csd->func_name = JS_NewString(ctx, func_name_str);
+    JS_FreeCString(ctx, func_name_str);
+    if (JS_IsException(csd->func_name))
+        csd->func_name = JS_NULL;
```

Comment:
Should likely either call `JS_GetException(ctx)` to clear the pending exception, or (preferably) return an error. Ditto on line 51718.

Reply:
Good call on glearing the exception. I was trying to avoid failing since there is no good way to report the error, is there?